datalad/distribution/install.py

```diff
@@ -565,196 +394,90 @@ def __call__(
-
-        lgr.debug("Installation of %s done.", current_dataset)
-
-        if not current_dataset.is_installed():
+        _handle_possible_annex_dataset(destination_dataset, reckless)
+
+        lgr.debug("Installation of %s done.", destination_dataset)
+
+        if not destination_dataset.is_installed():
             # log error and don't report as installed item, but don't raise,
             # since we might be in a process of recursive installation where
             # a lot of other datasets can still be installed successfully.
-            lgr.error("Installation of {0} failed.".format(current_dataset))
+            lgr.error("Installation of {0} failed.".format(destination_dataset))
         else:
-            installed_items.append(current_dataset)
+            installed_items.append(destination_dataset)
 
+        # we need to decrease the recursion limit, relative to
+        # subdatasets now
+        subds_recursion_limit = max(0, recursion_limit - 1) \
+                                  if isinstance(recursion_limit, int) \
+                                  else recursion_limit
         # Now, recursive calls:
-        if recursive and \
-                (recursion_limit is None
-                 or (recursion_limit and recursion_limit > 0)):
+        if recursive:
             if description:
                 lgr.warning("Description can't be assigned recursively.")
-            subs = [Dataset(p) for p in
-                    current_dataset.get_subdatasets(recursive=True,
-                                                    recursion_limit=1,
-                                                    absolute=True)]
-            for subds in subs:
-                try:
-                    rec_installed = Install.__call__(
-                        path=subds.path,
-                        dataset=current_dataset,
-                        recursive=True,
-                        recursion_limit=recursion_limit - 1
-                        if recursion_limit else None,
-                        if_dirty=if_dirty,
-                        save=save,
-                        git_opts=git_opts,
-                        git_clone_opts=git_clone_opts,
-                        annex_opts=annex_opts,
-                        annex_init_opts=annex_init_opts)
-                    if isinstance(rec_installed, list):
-                        installed_items.extend(rec_installed)
-                    else:
-                        installed_items.append(rec_installed)
-
-                except Exception:
-                    # Error itself should already be logged.
-                    lgr.info("{0} skipped.".format(subds))
+            subs = destination_dataset.get_subdatasets(
+                # yes, it does make sense to combine no recursion with
+                # recursion_limit: when the latter is 0 we get no subdatasets
+                # reported, otherwise we always get the 1st-level subs
```

Comment:
Re comment: Well, we could have the same result with `recursion=True` and `recursion_limit=0`. Furthermore: If the limit is 0, we wouldn't need to call that method at all. So, I think this possibility is more confusing than helpful. It means, that `recursion_limit` is respected even with no recursion (then: What about a higher limit but no recursion?) or to not respect it with no recursion except it is zero. Then we suddenly do respect it?

This actually is minor, but somehow relevant, since it's a matter of general style.

So, I'd prefer to see only `recursion=True` and `recursion_limit=0` working and otherwise check for 0 outside and don't even call `get_subdatasets()`.